rosidl_typesupport_opensplice_cpp/include/rosidl_typesupport_opensplice_cpp/requester.hpp

```diff
@@ -48,98 +53,173 @@ class Requester
-    assert(response_subscriber_ != nullptr);
-
-    std::string response_type_name = service_type_name_ + "_Response_";
-    std::string response_topic_name = service_name_ + "_Response";
+    if (!response_subscriber_) {
+      estr = "DomainParticipant::create_subscriber: failed for response";
+      goto fail;
+    }
 
-    // TODO(esteve): check that response_topic_ is valid
-    response_topic_ = participant->create_topic(
+    response_topic_ = participant_->create_topic(
       response_topic_name.c_str(), response_type_name.c_str(), default_topic_qos, NULL,
       DDS::STATUS_MASK_NONE);
-    assert(response_topic_ != nullptr);
-
-    DDS::StringSeq args;
-    args.length(2);
-    args[0] = DDS::string_dup(std::to_string(writer_guid_.first).c_str());
-    args[1] = DDS::string_dup(std::to_string(writer_guid_.second).c_str());
+    if (!response_topic_) {
+      estr = "DomainParticipant::create_topic: failed for response";
+      goto fail;
+    }
 
     // Let OpenSplice do any needed encoding
-    // TODO(esteve): check that content_filtered_response_topic_ is valid
-    content_filtered_response_topic_ = participant->create_contentfilteredtopic(
+    content_filtered_response_topic_ = participant_->create_contentfilteredtopic(
       service_name_.c_str(), response_topic_,
       "client_guid_0_ = %0 AND client_guid_1_ = %1",
       args);
-    assert(content_filtered_response_topic_ != nullptr);
+    if (!content_filtered_response_topic_) {
+      estr = "DomainParticipant::create_contentfilteredtopic: failed";
+      goto fail;
+    }
 
-    DDS::DataReaderQos default_datareader_qos;
-    response_subscriber_->get_default_datareader_qos(default_datareader_qos);
+    status = response_subscriber_->get_default_datareader_qos(default_datareader_qos);
+    if (nullptr != (estr = impl::check_get_default_datareader_qos(status))) {
+      goto fail;
+    }
 
-    // TODO(esteve): check that response_subscriber_ is valid
     response_datareader_ = response_subscriber_->create_datareader(
       content_filtered_response_topic_,
       default_datareader_qos, NULL, DDS::STATUS_MASK_NONE);
-    assert(response_datareader_ != nullptr);
+    if (!response_datareader_) {
+      estr = "Subscriber::create_datawriter: failed for response";
+      goto fail;
+    }
+    return nullptr;
+fail:
+    if (content_filtered_response_topic_) {
+      status = participant_->delete_contentfilteredtopic(content_filtered_response_topic_);
+      if (nullptr != impl::check_delete_contentfilteredtopic(status)) {
+        fprintf(stderr, "%s\n", impl::check_delete_contentfilteredtopic(status));
+      }
+    }
+    if (response_datareader_) {
+      // Assumption: subscriber is not null at this point.
+      status = response_subscriber_->delete_datareader(response_datareader_);
+      if (nullptr != impl::check_delete_datareader(status)) {
+        fprintf(stderr, "%s\n", impl::check_delete_datareader(status));
```

Comment:
Should it invoke the function once, check the return value and contionally use the return value for the error message? I can do those changes after merging the PR.